src/functions.php

```diff
@@ -231,6 +232,19 @@ function ($reason, $idx, Promise $aggregate) {
         ksort($results);
         return $results;
     });
+
+    if (true === $recursive) {
+        $promise = $promise->then(function ($results) use ($recursive, &$promises) {
+            foreach ($promises AS $promise) {
+                if (\GuzzleHttp\Promise\PromiseInterface::PENDING === $promise->getState()) {
+                    return all($promises, $recursive);
```

Comment:
Hello again,

I just checked this out, and all is fine: the wait function is not called again on a non-pending promise, so we can safely loop over the original iterator again without worrying about promises that may be settled multiple times.

I added that point into the unit tests to ensure that.